RiscVPkg/Universal/OpenSbi/Pei/OpenSbiPeim.c

```diff
@@ -0,0 +1,499 @@
+  0x02,  // TODO: Should be SBI spec version
+  PeiOpenSbiSetTimer,
+  PeiOpenSbiConsolePutChar,
+  PeiOpenSbiConsoleGetChar,
+  PeiOpenSbiClearIpi,
+  PeiOpenSbiSendIpi,
+  PeiOpenSbiRemoteFenceI,
+  PeiOpenSbiRemoteSfenceVma,
+  PeiOpenSbiRemoteSfenceVmaAsid,
+  PeiOpenSbiShutdown
+};
+
+EFI_PEI_PPI_DESCRIPTOR mOpenSbiLegacyPpiDesc = {
+  (EFI_PEI_PPI_DESCRIPTOR_PPI | EFI_PEI_PPI_DESCRIPTOR_TERMINATE_LIST),
+  &gPeiOpenSbiLegacyPpiGuid,
+  &mOpenSbiLegacyPpi
+};
+
+EFI_STATUS
+EFIAPI
+PeiOpenSbiHartStart (
+  IN  EFI_PEI_SERVICES             **PeiServices,
+  IN  PEI_RISCV_OPENSBI_LEGACY_PPI  *This
+  )
+{
+    // TODO
+    return EFI_SUCCESS;
+}
+
+EFI_STATUS
+EFIAPI
+PeiOpenSbiHartStop (
+  IN  EFI_PEI_SERVICES             **PeiServices,
+  IN  PEI_RISCV_OPENSBI_LEGACY_PPI  *This
+  )
+{
+    // TODO
+    return EFI_SUCCESS;
+}
+
+EFI_STATUS
+EFIAPI
+PeiOpenSbiHartGetStatus (
+  IN  EFI_PEI_SERVICES             **PeiServices,
+  IN  PEI_RISCV_OPENSBI_LEGACY_PPI  *This
+  )
+{
+    // TODO
+    return EFI_SUCCESS;
+}
+
+//
+// OpenSBI PPI interface function for the hart state management extension
+//
+PEI_RISCV_OPENSBI_HSM_PPI mOpenSbiHsmPpi = {
+  0x02,  // TODO: Should be SBI spec version
+  PeiOpenSbiHartStart,
+  PeiOpenSbiHartStop,
+  PeiOpenSbiHartGetStatus
+};
+
+EFI_PEI_PPI_DESCRIPTOR mOpenSbiHsmPpiDesc = {
+  (EFI_PEI_PPI_DESCRIPTOR_PPI | EFI_PEI_PPI_DESCRIPTOR_TERMINATE_LIST),
+  &gPeiOpenSbiHsmPpiGuid,
+  &mOpenSbiHsmPpi
+};
+
+EFI_STATUS
+EFIAPI
+PeiOpenSbiVendorCall (
+  IN  EFI_PEI_SERVICES             **PeiServices,
+  IN  PEI_RISCV_OPENSBI_VENDOR_PPI  *This,
+  IN  UINTN                          ExtensionId,
+  ...
+  )
+{
+    VA_LIST Args;
+    VA_START(Args, ExtensionId);
+
+    // TODO: Call SBI with all provided arguments
+
+    VA_END(Args);
+    return EFI_SUCCESS;
+}
+
+//
+// OpenSBI PPI interface function for the vendor extension
+//
+PEI_RISCV_OPENSBI_VENDOR_PPI mOpenSbiVendorPpi = {
+  0x00,
+  PeiOpenSbiVendorCall
+};
+
+EFI_PEI_PPI_DESCRIPTOR mOpenSbiVendorPpiDesc = {
+  (EFI_PEI_PPI_DESCRIPTOR_PPI | EFI_PEI_PPI_DESCRIPTOR_TERMINATE_LIST),
+  &gPeiOpenSbiVendorPpiGuid,
+  &mOpenSbiVendorPpi
+};
+
+EFI_STATUS
+EFIAPI
+PeiOpenSbiLibraryScratchThisHartPtr (
+  IN  EFI_PEI_SERVICES              **PeiServices,
+  IN  PEI_RISCV_OPENSBI_LIBRARY_PPI  *This,
+  OUT struct sbi_scratch            **ScratchPtr
+  )
+{
+    *ScratchPtr = (struct sbi_scratch *)csr_read(CSR_MSCRATCH);
+    // TODO: Can this ever fail?
+    return EFI_SUCCESS;
+}
+
+EFI_STATUS
+EFIAPI
+PeiOpenSbiLibraryPlatformPtr (
+  IN  EFI_PEI_SERVICES              **PeiServices,
+  IN  PEI_RISCV_OPENSBI_LIBRARY_PPI  *This,
+  IN  struct sbi_scratch             *ScratchPtr,
+  OUT const struct sbi_platform     **PlatformPtr
+  )
+{
+    if (ScratchPtr == NULL)
+        return EFI_INVALID_PARAMETER;
+
+    *PlatformPtr = (const struct sbi_platform *)(ScratchPtr->platform_addr);
+    // TODO: Do we need this function at all? In OpenSBI it's just a macro.
```

Comment:
cc @changab

Reply:
Should we maybe just put it in a header file?
`PeiOpenSbiLibraryScratchThisHartPtr` is also just a macro for inline asm.

Reply:
Seems could be :)